### ISSUE TITLE
FSDP optimizer not loaded correctly

### ISSUE THREAD

**Issue by howard-yen** [huggingface/transformers#25849]
### System Info

- `transformers` version: 4.33.0.dev0
- Platform: Linux-4.18.0-477.15.1.el8_8.x86_64-x86_64-with-glibc2.28
- Python version: 3.11.4
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.3
- Accelerate version: 0.22.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.0.1 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: yes, with FSPD 

### Who can help?

@pacman100 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. Use run_clm.py (https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_clm.py) to train a large model using the HuggingFace Trainer, use FSDP and save checkpoints. For example:

```
torchrun --nproc_per_node=4 --master_port=XXXXX experiments/run_clm.py \
    --model_name_or_path meta-llama/Llama-2-7b-hf \
    --dataset_name openwebtext \
    --streaming \
    --per_device_train_batch_size 16 \
    --gradient_accumulation_steps 1 \
    --do_train \
    --max_steps 1000 \
    --output_dir output_dir/ \
    --block_size 512 \
    --save_steps 10 \
    --save_total_limit 1 \
    --fsdp "full_shard auto_wrap" \
    --fsdp_transformer_layer_cls_to_wrap "LlamaDecoderLayer" \
    --bf16 True \
```

2.    Kill training after a checkpoint has been saved. Then, resume training from the checkpoint with the resume_from_checkpoint training argument.

4.   Observed behavior: optimizer state is not loaded

### Expected behavior

Optimizer state should be loaded. 

The current code checks if a scheduler was saved in the directory: 
https://github.com/huggingface/transformers/blob/07998ef39926b76d3f6667025535d0859eed61c3/src/transformers/trainer.py#L2444
but the current save_checkpoint method does not save the `self.lr_scheduler.state_dict()`:
https://github.com/huggingface/transformers/blob/07998ef39926b76d3f6667025535d0859eed61c3/src/transformers/trainer.py#L2334

Also, currently `accelerate.utils.save_fsdp_optimizer` saves the optimizer as `optimizer.bin`
https://github.com/huggingface/accelerate/blob/a16b843a1b8626d571e62f7301e3a478be2c04b4/src/accelerate/utils/fsdp_utils.py#L142C75-L142C75

but the trainer checks for `optimizer.pt` so `load_fsdp_optimizer` isn't called.
